File: src/main/java/dev/slne/discord/listeners/interaction/command/CommandReceivedListener.java
package dev.slne.discord.listeners.interaction.command;

import dev.slne.discord.DiscordBot;
import dev.slne.discord.interaction.command.DiscordCommand;
import dev.slne.discord.interaction.command.DiscordCommandManager;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import net.dv8tion.jda.api.interactions.commands.SlashCommandInteraction;

public class CommandReceivedListener extends ListenerAdapter {

    @Override
    public void onSlashCommandInteraction(SlashCommandInteractionEvent event) {
        SlashCommandInteraction interaction = event.getInteraction();

        DiscordCommandManager commandManager = DiscordBot.getInstance().getCommandManager();
        DiscordCommand command = commandManager.findCommand(interaction.getName());

        if (command != null) {
            command.execute(event);
        }
    }

}

File: src/main/kotlin/dev/slne/discordold/listener/interaction/command/CommandReceivedListener.kt
package dev.slne.discordold.listener.interaction.command

import dev.minn.jda.ktx.events.listener
import dev.slne.discordold.discord.interaction.command.DiscordCommandProcessor
import jakarta.annotation.PostConstruct
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import org.springframework.stereotype.Component

@Component
class CommandReceivedListener(
    private val jda: JDA,
    private val processor: DiscordCommandProcessor
) {

    @PostConstruct
    fun registerListener() {
        jda.listener<SlashCommandInteractionEvent> { event ->
            processor.getCommand(event.name)?.second?.execute(event)
                ?: error("Command ${event.name} not found")
        }
    }
}
